socks/socks4.py

```diff
@@ -67,11 +88,12 @@ def request(
         elif address_type == AddressType.DN and not self.allow_domain_names:
             raise SOCKSError("Domain names only supported by SOCKS4A")
 
-        raise NotImplementedError()
+        request = SOCKS4Request(command, port, encoded_addr, user_id)
+        self._data_to_send += request.dumps()
 
-    def receive_data(self, data: bytes) -> typing.List[SOCKS4Reply]:
+    def receive_data(self, data: bytes) -> SOCKS4Reply:
         self._received_data += data
-        raise NotImplementedError()
+        return SOCKS4Reply.loads(self._received_data)
```

Comment:
We need to be able to handle multiple replies since you can send multiple requests concurrently so the type annotation should stay the same

Also we should only send in 8 bytes at a time to `SOCKS4Reply.loads()`

Reply:
> We need to be able to handle multiple replies since you can send multiple requests concurrently

Sorry, I think I'm missing something. Are we expecting to send multiple SOCKS requests to the *same* proxy server?

Reply:
You're totally right! Sorry about that :)